### PR TITLE
[Process] Introduce InputStream and iterator for output

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | 3.1 |
| Fixed tickets | - |

### CODE REVIEW

File: components/process.rst [hunk 1]
```diff
@@ -128,6 +128,20 @@ are done doing other stuff::
...
+Before a process is started, you can specify its standard input using either the
+:method:`Symfony\\Component\\Process\\Process::setInput` method or the 4th argument
+of the Process constructor.
+
+The provided input can be a string, a stream resource or a Traversable object. When
+this input is fully written to the subprocess standard input, the corresponding pipe
+is closed.
+
+In order to write to a subprocess standard input while it is running, the component
+provides the :class:`Symfony\\Component\\Process\\InputStream` class.
```
Comment (stof) [inline components/process.rst]: documenting its usage is necessary

Comment (fabpot) [inline components/process.rst]: I was about the say the same. A small practical example would help a lot.

**Comment by stof** [comment]
I think it might be time to split the doc of the Process component into multiple chapters, an dhaving one such chapter dedicated to the way to deal with the input (with basic examples and more complex ones)

**Comment by iltar** [comment]
I think @stof has a valid point here, the process page is already rather big compared to some others.

**Comment by fabpot** [comment]
Don't forget to also document "IteratorAggregate to stream output" from symfony/symfony#18414

**Comment by nicolas-grekas** [comment]
PR updated, LGTM. Do you think it needs more content?
Comments welcomed. If someone is willing to help and do the split @stof suggests in a later PR, I'd be thankful for it...

File: components/process.rst [hunk 2]
```diff
@@ -43,13 +43,28 @@ The ``getOutput()`` method always returns the whole content of the standard
...
+methods return the new outputs since their last call.
 
 The :method:`Symfony\\Component\\Process\\Process::clearOutput` method clears
 the contents of the output and
 :method:`Symfony\\Component\\Process\\Process::clearErrorOutput` clears
 the contents of the error output.
 
+You can also use the :class:`Symfony\\Component\\Process\\Process` class with the
+foreach construct to get the output while it is generated. By default, the loop waits
+for new outputs before going to the next iterations::
```
Comment (fabpot) [inline components/process.rst]: output, iteration?

**Comment by fabpot** [comment]
:+1:

File: components/process.rst [hunk 3]
```diff
@@ -128,6 +143,46 @@ are done doing other stuff::
...
+    $input->write('bar');
+    $input->close();
+
+    $process->wait();
+
+    // will echo: foobar
+    echo $process->getOutput();
+
+The :method:`Symfony\\Component\\Process\\InputStream::write` method accepts scalars,
+stream resources or Traversable objects as argument. As shown is the above example,
```
Comment (javiereguiluz) [inline components/process.rst]: `As shown is the` -> `As shown in the`

File: components/process.rst [hunk 4]
```diff
@@ -128,6 +143,46 @@ are done doing other stuff::
     which means that your code will halt at this line until the external
     process is completed.
 
+Feeding the standard input of a Process
```
Comment (javiereguiluz) [inline components/process.rst]: `Streaming` instead of `Feeding` ?

**Comment by javiereguiluz** [comment]
:+1:

@nicolas-grekas thanks for these docs!

File: components/process.rst [hunk 5]
```diff
@@ -43,13 +43,28 @@ The ``getOutput()`` method always returns the whole content of the standard
 output of the command and ``getErrorOutput()`` the content of the error
 output. Alternatively, the :method:`Symfony\\Component\\Process\\Process::getIncrementalOutput`
 and :method:`Symfony\\Component\\Process\\Process::getIncrementalErrorOutput`
-methods returns the new outputs since the last call.
+methods return the new output since their last call.
```
Comment (xabbuh) [inline components/process.rst]: We should make this fix in all affected lower branches too.

File: components/process.rst (hunk 4 above)
Comment (xabbuh) [inline components/process.rst]: "Feeding" sounds good to me.

Comment (xabbuh) [inline components/process.rst]: "Standard" and "Input" should be uppercased.

**Comment by xabbuh** [comment]
We need a `versionadded` directive for the features that are new in Symfony 3.1.

File: components/process.rst [hunk 6]
```diff
@@ -128,6 +143,46 @@ are done doing other stuff::
     which means that your code will halt at this line until the external
     process is completed.
 
+Feeding the standard input of a Process
+---------------------------------------
+
+Before a process is started, you can specify its standard input using either the
+:method:`Symfony\\Component\\Process\\Process::setInput` method or the 4th argument
+of the Process constructor. The provided input can be a string, a stream resource or
```
Comment (WouterJ) [inline components/process.rst]: "of the constructor."

**Comment by WouterJ** [comment]
Hi @nicolas-grekas! Can you please finish this PR? Otherwise, just say it and we'll fix the minor issues while merging.

**Comment by nicolas-grekas** [comment]
Comments addressed hopefully (I'll let you guys split this into several pages...)

File: components/process.rst (hunk 5 above)
Comment (xabbuh) [inline components/process.rst]: see #6658

File: components/process.rst [hunk 7]
```diff
@@ -68,6 +83,10 @@ with a non-zero code)::
         echo $e->getMessage();
     }
 
+.. versionadded:: 3.1
+    Streaming the input or the output of a process using iterators
+    were added in Symfony 3.1
```
Comment (xabbuh) [inline components/process.rst]: I think we should split this.

Above the paragraph added above:

```rst
.. versionadded:: 3.1
    Support for streaming the output using iterators was introduced
    in Symfony 3.1.
```

After the "Streaming to the standard input of a Process" headline:

```rst
    Support for streaming the input of a process was introduced in
    Symfony 3.1.
```

File: components/process.rst [hunk 8]
```diff
@@ -128,6 +147,46 @@ are done doing other stuff::
     which means that your code will halt at this line until the external
     process is completed.
 
+Streaming to the standard input of a Process
```
Comment (xabbuh) [inline components/process.rst]: Streaming to the Standard Input of a Process

**Comment by xabbuh** [comment]
👍 except of the minor comments above (but we can address them while merging)

Status: Reviewed

**Comment by xabbuh** [comment]
ping @javiereguiluz @weaverryan @WouterJ

**Comment by WouterJ** [comment]
Thanks @nicolas-grekas! Fyi, I've rebased this PR in https://github.com/symfony/symfony-docs/pull/6705 and merged that one.

**Pull request closed without merging**